src/main/java/seedu/address/model/leave/Date.java

```diff
@@ -0,0 +1,69 @@
+package seedu.address.model.leave;
+
+import static seedu.address.commons.util.CollectionUtil.requireAllNonNull;
+
+import java.time.LocalDate;
+
+/**
+ * Represents a Date in the address book.
+ * Guarantees: immutable; is valid as declared in {@link #isValidDate(String)}
+ */
+public class Date {
+
+    private static final String MESSAGE_CONSTRAINTS = "Date should be in the format YYYY-MM-DD";
+    private final LocalDate date;
+
+    private Date(LocalDate date) {
+        this.date = date;
+    }
+
+    /**
+     * Factory method for constructing a Date object with a LocalDate object.
+     *
+     * @param date
+     * @return Date object
+     */
+    public static Date of(LocalDate date) {
+        requireAllNonNull(date);
+        return new Date(date);
+    }
+
+    /**
+     * Factory method for constructing a Date object with a String.
+     *
+     * @param date
+     * @return Date object
+     */
+    public static Date of(String date) throws IllegalArgumentException {
+        requireAllNonNull(date);
+        if (!date.matches("\\d{4}-\\d{2}-\\d{2}")) {
```

Comment:
LocalDate.parse() already checks for this; you can wrap LocalData.parse in a try-catch block and then catch the DateTimeException. Also, it might be a good idea to set up the DateTimeFormatter to resolve dates strictly i.e. https://github.com/foal/gwt-time/issues/10, so we will not have to manually check for illegal dates e.g. September 31